### DISCUSSION TITLE
Can't use setHtmlAttribute with uno.winui.

### DISCUSSION THREAD

**Discussion by electriceop3** [unoplatform/uno#12616] [Q&A] [answered]
I am needing to use this.setHtmlAttribute. I have installed Uno.WinUi.Runtime.Webassembly as it said at this link: [Embedding Existing JavaScript Components Into Uno-WASM](https://platform.uno/docs/articles/interop/wasm-javascript-1.html), but when I try to setHtmlAttribute, I get an error saying The type 'UIElement' is defined in an assembly that is not referenced. You must add a reference to assembly 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'. First of all, I can't find a version 255.255.255.255, but when I install the latest version, 4.9.20, it says Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced., and even after deleting any Uno.Ui package references, that error persists until I create a new Uno project and move any files to that. Here is the code that was throwing an error:
```
using Microsoft.UI.Xaml.Controls;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Uno.UI.Runtime.WebAssembly; // when I add [HtmlElement("div")] after installing Uno.Winui.Runtime.WebAssembly it automatically adds this. It doesn't throw any errors on its own, only when I add setHtmlAttribute
using Windows.System.Preview;

namespace Project.XamlElements
{
    [HtmlElement("div")]
    public sealed partial class MapControl: FrameworkElement
    {
        public MapControl()
        {
          #if __WASM__ 
            //has error with or without conditional
            
            this.SetHtmlAttribute("id", "4");
          #endif
        }
    }
}
```
I think the problem might be that I need to make the package reference conditional, but I can't find anything on the docs for what the targetName for wasm is.

```
//<ItemGroup Condition = "'$(TargetFramework)' == 'wasm? or webassembly?' ">
    //<PackageReference Include="Uno.WinUI.Runtime.WebAssembly" />
//</ItemGroup>
```

**Comment by jeromelaban** [comment, marked as answer]
Thank for the report. The documentation is slightly out of date with regards to the use of external components and we're expecting it to change with the upcoming previews of .NET 8 with the introduction of `net8.0-browser`.

If your project only targets WebAssembly, then using conditionals as you mentioned (using `net7.0`) will be enough. If you target Skia in addition to WebAssembly, you'll also need to include some runtime validation to determine if the current environment is webassembly or not, like this:

https://github.com/unoplatform/uno/blob/101f43a1fd305f00c84f1826ff6eaa9977884489/src/Uno.Foundation.Runtime.WebAssembly/Interop/PlatformHelper.cs#L51-L54

**Reply by electriceop3** [reply]
Which conditionals were you referring to for .net 7, `#if` and `#endif` or `<ItemGroup Condition = "'$(TargetFramework)' == ''target framework">
</ItemGroup>`?

**Reply by jeromelaban** [reply]
You'll need to use both conditions. In a class library, `__WASM__` will not work (not until `net8.0-browser`), so you'll need to add a `#if NET7_0_OR_GREATER` to make the block `net7.0` specific, and use `Condition="'$(TargetFramework)' == ''net7.0">` in the csproj.

**Reply by electriceop3** [reply]
Okay that got rid of the error. Thank you!

**Reply by electriceop3** [reply]
Just curious, is there a way to setHtmlElement of an existing xaml element?

**Reply by jeromelaban** [reply]
@electriceop3 it's not possible, that attribute is read immediately in the constructor. What would be the scenarion that you're looking for?

**Reply by electriceop3** [reply]
I was just wondering if I could use the xaml elements in Microsoft.UI.Xaml.Controls or Microsoft.UI.Xaml.Panels instead of defining my own that would have the same basic functionality.

**Reply by jeromelaban** [reply]
Thanks. What do you need to be present on all controls?

**Reply by electriceop3** [reply]
I'm not sure that I understood your question correctly, but I don't have anything that needs to be on all controls per say, there are just some controls in my code that I would like to be able to assign an id so that it would work with JavaScript, and it to me it seems a little redundant to code a separate class every time I want to assign a specific id to a control or panel, particularly when the elements would have the same basic functionality as the elements in Microsoft.UI.Xaml.Controls or Microsoft.UI.Xaml.Panels.

**Reply by jeromelaban** [reply]
I see, thanks. For uno controls, the ID is handled exclusively by Uno (it contains a managed handle id) and it should not be modified. If you need discoverable IDs, you'll need to manually create children and add the ID there.

**Reply by electriceop3** [reply]
Okay, thanks! Do you know if there is a way to supply a xaml element with a parameter, so that every instance that needs a unique id doesn't need its own class and I can instead just pass the id as a parameter?

**Reply by jeromelaban** [reply]
I'm not sure to follow the ask. Do you have an example of the API you wish to see ?

**Reply by electriceop3** [reply]
I was just wondering if you could pass parameters or arguments to a xaml instance, but I found something: [Pass Arguments](https://learn.microsoft.com/en-us/dotnet/maui/xaml/pass-arguments). Thank you for your help!

**Reply by jeromelaban** [reply]
Got it. Note that the documentation you've linked to is for MAUI, and at this time for WinUI and Uno there's no way to provide constructor parameters to a type. If you wish to do so, you'll need to do it through C# code.

**Reply by electriceop3** [reply]
Okay, thanks! I also had found this thread on StackOverflow: [Calling a parameterized constructor from XAML](https://stackoverflow.com/questions/1083159/calling-a-parameterized-constructor-from-xaml), but apparently, according to a linked StackOverflow thread: [How to fix: XAML2009 language construct is not allowed here?](https://stackoverflow.com/questions/14347181/how-to-fix-xaml2009-language-construct-is-not-allowed-here), it doesn't actually work. Sorry for the trouble, I'm not very experienced with xaml.

**Comment by BertrandDeSaintBezier** [comment]
Bumping this as I have the exact same issue but adding the following conditional statement in the shared .csproj
```xaml
<ItemGroup Condition="'$(TargetFramework)' == 'net7.0'">
	<PackageReference Include="Uno.WinUI.Runtime.WebAssembly" Version="4.9.26" />	
</ItemGroup>
```
and #if statements around the relevant lines of code
```csharp
...
#if NET7_0_OR_GREATER
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Create("BROWSER")))
    {
        this.SetCssStyle(string.Empty, string.Empty);
        this.ExecuteJavascript(string.Empty);
    }
#endif
...
```
results in different issues : `'TestUserControl' does not contain a definition for 'ExecuteJavascript' and no accessible extension method 'ExecuteJavascript' accepting a first argument of type 'TestUserControl' could be found (are you missing a using directive or an assembly reference?)     Project : [ApplicationName] (net7.0-windows10.0.19041)` (same error message for `SetCssStyle`).

I tried using these methods in a UserControl, I don't know if this is supported or not (I believe that UserControls inherit from FrameworkElements ?).

EDIT : I should perhaps also specify that this specific project targets Windows & WASM.

**Reply by jeromelaban** [reply]
For this particular sample, `NET7_0_OR_GREATER` is making the block valid for windows as well (since it's net7.0-windows). You'll need to change your conditional like this:

```csharp
#if HAS_UNO
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Create("BROWSER")))
    {
        this.SetCssStyle(string.Empty, string.Empty);
        this.ExecuteJavascript(string.Empty);
    }
#endif
```

Considering that you're targeting windows and Wasm only, this should be enough to get you going. We're also in the process of updating the documentation for those scenarios, this should get somehow clearer soon.

**Reply by BertrandDeSaintBezier** [reply]
Okay I see. This actually reminded me of [another thread I had started](https://github.com/unoplatform/uno/discussions/12313#discussioncomment-5947765) where you mentionned that when building for Windows, Uno just delegates to WinUI instread of running its own code. Makes a lot more sense now seeing it in action. Also clears up how the `#HAS_UNO` conditional works/what it's useful for.

Am I incorrect in assuming that `#HAS_UNO_WASM` doesn't quite work yet ?

Regardless, thanks for yet another quality and speedy answer ! 😄

**Reply by jeromelaban** [reply]
Thanks!!

For the `HAS_UNO_WASM` define, it's the same as the original comment, which is that it cannot be defined until `net8.0-browser` becomes available (at least when working in a cross target library).
